### PR TITLE
Create location input component

### LINKED ISSUE #7: Create location input component
## Description

To allow users to input their precise location, we will use a combination of Mapbox's [GeolocateControl](https://docs.mapbox.com/mapbox-gl-js/example/locate-user/), [Draggable Marker](https://docs.mapbox.com/mapbox-gl-js/example/drag-a-marker/), and [Geocoder Input](https://docs.mapbox.com/mapbox-gl-js/example/mapbox-gl-geocoder-outside-the-map/). Users should be able to: 
- Locate their exact location and then specify the exact location of their solar panels by dragging a marker 
- Enter an address and then specify the exact location of their solar panels by dragging a marker
- Simply drag/zoom the map and then specify the exact location of their solar panels by dragging a marker

Mapbox is very versatile, and our solution should be iterated upon to find the best fit. 

## Todos

These points are a rough guideline. **Please** feel free to discuss with others on the team about the best way to design this component!

- [x] acquire mapbox access token (we should be able to use free tier for now) 
- [x] create map component 
- [ ] implement geocoding input 
- [x] implement placable marker 
- [ ] ensure mobile-friendliness 
- [ ] style according to designs (future task—may still be blocked upon completion of all tasks above)

## Useful links

- [tutorial for mapbox with react](https://docs.mapbox.com/help/tutorials/use-mapbox-gl-js-with-react/)
- [geocoder with coordinates!](https://docs.mapbox.com/mapbox-gl-js/example/mapbox-gl-geocoder-accept-coordinates/)
- [draggable marker](https://docs.mapbox.com/mapbox-gl-js/example/drag-a-marker/)
- [locate user](https://docs.mapbox.com/mapbox-gl-js/example/locate-user/)

## Time estimate
~ 2 weeks

### PR DESCRIPTION
**Pull request by jackypark9852**
## Description

Fixes #<number>

### CODE REVIEW

**Comment by AndrewLester** [comment]
@all-contributors please add @jackypark9852 for code

**Comment by AndrewLester** [comment]
@all-contributors please add @ashayp22 for code

File: components/Button.tsx [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import React, { ReactNode } from 'react';
+
+interface Props {
+  children: ReactNode;
```
Comment (AndrewLester) [inline components/Button.tsx]: The React `PropsWithChildren` will cover this for ya

```suggestion

```

File: components/Button.tsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import React, { ReactNode } from 'react';
```
Comment (AndrewLester) [inline components/Button.tsx]: Next.js makes it so `import React from 'react';` isn't required, pretty cool. Also, `PropsWithChildren` covers `ReactNode` as you'll see below.

```suggestion
import { FC, PropsWithChildren } from 'react';
```

File: components/Button.tsx [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+interface Props {
+  children: ReactNode;
+  enabled: boolean;
+}
+
+export default function Button({ children, enabled }: Props) {
```
Comment (AndrewLester) [inline components/Button.tsx]: Using an arrow function + `FC` type will make this type definition a bit simpler
```suggestion
const Button: FC<PropsWithChildren<Props>> = ({ children, enabled }) => {
```

File: components/Button.tsx [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  } else {
+    return (
+      <button
+        disabled
+        className="bg-ocf-gray dark:bg-ocf-gray shadow h-full w-full text-center rounded-md font-bold text-xl uppercase"
+      >
+        {children}
+      </button>
+    );
+  }
```
Comment (AndrewLester) [inline components/Button.tsx]: This if can be translated into `disabled={!enabled}`. When you think about it, maybe `disabled={disabled}` could  be simpler? What do you think?
```suggestion
    return (
      <button
        disabled={!enabled}
        className="bg-ocf-gray dark:bg-ocf-gray shadow h-full w-full text-center rounded-md font-bold text-xl uppercase"
      >
        {children}
      </button>
    );
```

File: components/mapbox_input.tsx [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  setIsSubmissionEnabled: (isSubmissionEnabled: boolean) => void;
+  zoomLevelThreshold: number;
+}
+
+export default function MapBoxInput({
+  setIsSubmissionEnabled,
+  zoomLevelThreshold,
+}: MapBoxInputProps) {
+  mapboxgl.accessToken =
+    'pk.eyJ1IjoiYWxlc3RlcjMiLCJhIjoiY2xlM3JwdDkwMDR6cjNvdGRpanZqZHd0ciJ9.ibQNGDwEE_Wc59LB2dhs9Q';
```
Comment (AndrewLester) [inline components/mapbox_input.tsx]: We'll probably want to store this in an environment variable, despite it being public and fine to expose. You can put this in your `.env` and access it with `process.env.<ENV_VAR_NAME>`

File: components/mapbox_input.tsx [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+interface MapBoxInputProps {
+  setIsSubmissionEnabled: (isSubmissionEnabled: boolean) => void;
+  zoomLevelThreshold: number;
+}
+
+export default function MapBoxInput({
+  setIsSubmissionEnabled,
+  zoomLevelThreshold,
+}: MapBoxInputProps) {
```
Comment (AndrewLester) [inline components/mapbox_input.tsx]: Swap this to an arrow function and use the `FC` type like above.

File: components/mapbox_input.tsx [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+  return (
+    <div className="flex flex-col h-full">
+      <div ref={geocoderContainer} id="geocoderContainer" />
+      <div ref={mapContainer} className="h-screen" />
+    </div>
+  );
+}
+
+function UpdateMarker(
```
Comment (AndrewLester) [inline components/mapbox_input.tsx]: Camel case

```suggestion
function updateMarker(
```

File: package.json [hunk 8]
```diff
@@ -12,13 +12,18 @@
     "style": "yarn format && yarn lint"
   },
   "dependencies": {
+    "@mapbox/mapbox-gl-geocoder": "^5.0.1",
+    "@next/font": "13.1.6",
```
Comment (AndrewLester) [inline package.json]: Sadly we have to remove this since we downgraded next.js :(
```suggestion

```

File: pages/form.tsx [hunk 9]
```diff
@@ -1,6 +1,6 @@
 import Input from '~/components/Input';
 import { withPageAuthRequired } from '~/lib/auth';
-
+import Button from '~/components/Button';
```
Comment (AndrewLester) [inline pages/form.tsx]:
```suggestion
import Button from '~/components/Button';

```

**Review by AndrewLester — CHANGES_REQUESTED** [review]
Great work on this! Lots of code in here and most of it is perfect. Just a few formatting/organization things really, also you can remove `package-lock.json` since our `yarn.lock` makes up for it.

**Review by AndrewLester — APPROVED** [review]

**Review by AndrewLester — APPROVED** [review]

File: pages/test/location_input.tsx [hunk 10]
```diff
@@ -0,0 +1,28 @@
+import { useState } from 'react';
+import Button from '~/components/Button';
+import MapBoxInput from '~/components/mapbox_input';
+
+export default function Home() {
+  const [isSubmissionEnabled, setIsSubmissionEnabled] = useState(false);
+  const [zoomLevelThreshold, setZoomLevelThreshold] = useState(10);
```
Comment (ericcccsliu) [inline pages/test/location_input.tsx]: Don't really get why this is in a useState—seems like this is something we can just hardcode

File: components/Button.tsx [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+interface Props {
+  enabled: boolean;
+}
+
+const Button: FC<PropsWithChildren<Props>> = ({ children, enabled }) => {
+  return (
+    <button
+      disabled={!enabled}
+      className="bg-ocf-yellow disabled:bg-ocf-gray disabled:dark:bg-ocf-gray shadow h-full w-full text-center rounded-md font-bold text-xl uppercase"
```
Comment (ericcccsliu) [inline components/Button.tsx]: We can probably set a specific height on this button component—perhaps `h-14` could work? I don't see the height of this component changing all that much, if at all

Comment (ericcccsliu) [reply components/Button.tsx]: also maybe consider a max width—currently the button looks a big wonky as screen sizes get larger

**Review by ericcccsliu — APPROVED** [review]
Looks amazing!!!!! left a few very tiny code quality suggestions. Awesome to see this map input come together—I think you absolutely knocked it out of the park with this one!!!!!!!!

**Review by AndrewLester — APPROVED** [review]

**Review by ericcccsliu — APPROVED** [review]
LGTM

**Pull request merged**